esp-config/src/generate/mod.rs

```diff
@@ -45,8 +46,466 @@ impl fmt::Display for Error {
+    /// Condition which makes this default value used.
+    /// You can and have to have exactly one active default value.
+    #[serde(rename = "if")]
+    #[serde(default = "true_default")]
+    pub if_: String,
+    /// The default value.
+    pub value: Value,
+}
+
+/// A configuration option.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(deny_unknown_fields)]
+pub struct CfgOption {
+    /// Name of the configuration option
+    pub name: String,
+    /// Description of the configuration option.
+    /// This will be visible in the documentation and in the tooling.
+    pub description: String,
+    /// A condition which specified when this option is active.
+    #[serde(default = "true_default")]
+    pub active: String,
+    /// The default value.
+    /// Exactly one of the items needs to be active at any time.
+    pub default: Vec<CfgDefaultValue>,
+    /// Constraints (Validators) to use.
+    /// If given at most one item is allowed to be active at any time.
+    pub constraints: Option<Vec<CfgConstraint>>,
+    /// A display hint for the value.
+    /// This is meant for tooling and/or documentation.
+    pub display_hint: Option<DisplayHint>,
+    /// The stability guarantees of this option.
+    #[serde(default = "unstable_default")]
+    pub stability: Stability,
+}
+
+/// A conditional constraint / validator for a config option.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(deny_unknown_fields)]
+pub struct CfgConstraint {
+    /// Condition which makes this validator used.
+    #[serde(rename = "if")]
+    #[serde(default = "true_default")]
+    if_: String,
+    /// The validator to be used.
+    #[serde(rename = "type")]
+    type_: Validator,
+}
+
+/// Generate the config from a YAML definition.
+///
+/// The YAML follows the format outlined by [Config].
+///
+/// After deserializing the config and normalizing it, this will call
+/// [generate_config] to finally get the currently active configuration.
+pub fn generate_config_from_yaml_definition(
+    yaml: &str,
+    enable_unstable: bool,
+    emit_md_tables: bool,
+    chip: Option<esp_metadata::Config>,
+) -> Result<HashMap<String, Value>, Error> {
+    #[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash)]
+    pub struct I128NumericTypes;
+
+    #[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash)]
+    pub struct I128(pub i128);
+
+    impl FromStr for I128 {
+        type Err = core::num::ParseIntError;
+
+        fn from_str(s: &str) -> Result<Self, Self::Err> {
+            Ok(I128(s.parse()?))
+        }
+    }
+
+    impl core::fmt::Display for I128 {
+        fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+            write!(f, "{}", self.0)
+        }
+    }
+
+    use evalexpr::{EvalexprError, EvalexprResult};
+
+    impl<NumericTypes: evalexpr::EvalexprNumericTypes<Int = Self>>
+        evalexpr::EvalexprInt<NumericTypes> for I128
```

Comment:
I'm only commenting this because of other changes I'd like you to make, but it's very awkward to read around a type I don't care about in a function that would be otherwise ~50loc. Could you please move this to a separate file, like `evalexpr_extensions.rs`?